### ISSUE TITLE
Base64 encoding can only elide padding when the size of encoded data is known

### ISSUE THREAD

**Issue by peterbourgon** [Cyphrme/Coze#17]
https://github.com/Cyphrme/Coze/blob/01c154e4024b4e876b8d152166ce85cf2a945e22/README.md#coze-fields

> Binary values are encoded as RFC 4648 base64 URI with padding truncated (b64ut). 

https://www.rfc-editor.org/rfc/rfc4648#section-3.2

> when assumptions about the size of transported data cannot be made, padding is required to yield correct decoded data.

As far as I can tell, the size of binary values is not communicated to recipients, and therefore padding should not be truncated. (The URI encoding is also non-standard.)

**Comment by zamicol** [comment]
How is the URI encoding non-standard?  

---

There's some general things about RFC base64 that should be said first: Padding
characters help satisfy length requirements and carry no other meaning. It's
always possible to determine the length of the input unambiguously from the
length of the encoded sequence. Since Coze does not concatenating unpadded
base64 strings, (Coze only concatenate in the binary form, not the base64 form)
Coze does not need base64 padding.  For RFC base64 [/w== and /w are equal to
11111111](https://convert.zamicol.com/#?inAlph=ub64p&in=%252Fw%253D%253D&outAlph=01)
(I'm pretty fluent with base conversion and I build that tool) 


As a historical anecdote, Coze used to encode with Hex, because it is more human
readable and because Hex is always twice as large as the binary form, while
b64ut does not have a static multiplicative relationship with binary and is less
human readable. After considering the message size savings, we dropped Hex in
favor of RFC b64ut.  Satoshi had the same concerns with base64 and thus base 58.
We decided that RFC base64 is good enough and that implementing an alternative
base conversion system would be more trouble than it's worth.  That has not
stopped other from doing so (See Keybase's solution, it's linked at the bottom
of the base conversion tool along with others)

As another aside, [JOSE already does the same
thing](https://www.rfc-editor.org/rfc/rfc7515#appendix-C), and it's widely used
in industry.

To summarize, there's a few reasons why padding is not needed:

0. The purpose of padding is to explicitly denote "empty bytes".  It's already
   redundant.
1. It's in JSON.
2. Digests and Cryptographic signatures. 


Padding is never needed to correctly decode a base64 message as long as the
whole message was transported.  
Coze also doesn't have to worry about base64 concatenation since base64 is in
JSON.  Double quote serves as the base64 string terminator.  
 If something went wrong with encoding or transport, Coze verification simply
 fails since cryptographic functions are serving the function of integrity
 checking.  As long as a Coze message verifies, it's also integral.

**Comment by zamicol** [comment]
I don't mean "close the issue" for no more feedback, but I don't believe this is a concern.  (I'm a bit of a GitHub dunce, please forgive any of my social blunders.)  

I appreciate you reading Coze and poking holes into it.  This is exactly what needs to be done, and I want to motivate skepticism as much as I can.

**Comment by peterbourgon** [comment]
> How is the URI encoding non-standard?

[You use base64.URLEncoding](https://github.com/Cyphrme/Coze/blob/01c154e4024b4e876b8d152166ce85cf2a945e22/base64.go#L20). That [is described](https://pkg.go.dev/encoding/base64#pkg-variables) as 

> URLEncoding is **the alternate base64 encoding** defined in [RFC 4648](https://rfc-editor.org/rfc/rfc4648.html). It is typically used in URLs and file names.

whereas base64.StdEncoding is described as

> StdEncoding is **the standard base64 encoding**, as defined in [RFC 4648](https://rfc-editor.org/rfc/rfc4648.html).

---

> Padding characters help satisfy length requirements and carry no other meaning. It's always possible to determine the length of the input unambiguously from the length of the encoded sequence. 

As far as I can tell, this is a mis-reading of the relevant requirements, and not correct. It is only possible to unambiguously decode a base64 encoded string in isolation if padding characters are included. If padding characters are elided, then it is only possible to unambiguously decode that string if the length is communicated out-of-band.

Quoting https://www.rfc-editor.org/rfc/rfc4648#section-3.2

```
   In some circumstances, the use of padding ("=") in base-encoded data
   is not required or used.  In the general case, when assumptions about
   the size of transported data cannot be made, padding is required to
   yield correct decoded data.

   Implementations MUST include appropriate pad characters at the end of
   encoded data unless the specification referring to this document
   explicitly states otherwise.
```

**Comment by zamicol** [comment]
> base64.StdEncoding

It's a matter of semantics.  base64.URLEncoding is standardized by the same RFC.  We've dubbed it more specifically b64ut.  Even though it's not what the RFC names as the standard alphabet, URI encoding is standardized formally by that RFC 4648.  

We especially felt the need to dub it b64ut to avoid confusion with the [generalized arbitrary base 64](https://convert.zamicol.com/#?inAlph=Text&in=Hello%2520Peter!&outAlph=ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_) which uses the "iterative divide by radix" method and is sometimes equal to RFC base64.  


---


The JOSE JWS RFC specifically addresses that:

 >As per the example code above, the number of '=' padding characters
   that needs to be added to the end of a base64url-encoded string
   without padding to turn it into one with padding is a deterministic
   function of the length of the encoded string.  Specifically, if the
   length mod 4 is 0, no padding is added; if the length mod 4 is 2, two
   '=' padding characters are added; if the length mod 4 is 3, one '='
   padding character is added; if the length mod 4 is 1, the input is
   malformed.

And that is correct.  Padding is always deterministically recreatable as long as the original message is given.  


   
 RFC 4648 is basically referring to streaming, where the original message may _not_ be given, and I think it's one of the more confusingly worded section.  If a stream ends mid stream, without padding it may not be known that the stream ended or if there's an error.  For batch processing, this isn't relevant.  Firstly, the transport (TCP) will most likely error, then JSON itself will be malformed, the digest will be bad, and the cryptographic signature will not be valid.  There's many layers of defense against having to worry about padding in Coze.

**Comment by peterbourgon** [comment]
> It's a matter of semantics. base64.URLEncoding is standardized by the same RFC. We've dubbed it more specifically b64ut. Even though it's not what the RFC names as the standard alphabet, URI encoding is standardized formally by that RFC 4648.

"Standard" does not mean "any one of the alphabets defined by the authoritative RFC", it means "the specific alphabet denominated as Standard by the authoritative RFC", which is explicitly not the URI encoding.

> Padding is always deterministically recreatable as long as the original message is given . . .  RFC 4648 is basically referring to streaming, where the original message may not be given, and I think it's one of the more confusingly worded section. If a stream ends mid stream, without padding it may not be known that the stream ended or if there's an error.

None of these claims are correct. Padding is not deterministically re-createable, because "the original message" is not guaranteed to be knowable by a given recipient. Neither does RFC 4648 apply only to "streaming" use cases.

I'll disengage at this point.

**Comment by zamicol** [comment]
I opened a new issue that's related to base 64 encoding:  https://github.com/Cyphrme/Coze/issues/18
